File: Graph/python/bfs.py
from collections import deque
from typing import List

from .graph_utils import Graph


def breadth_first_traversal(graph: Graph, source: str) -> List[str]:
    if not source:
        return []

    order: List[str] = []
    queue: deque[str] = deque([source])
    visited = {source}

    while queue:
        node = queue.popleft()
        order.append(node)

        for neighbor in graph.get(node, []):
            if neighbor in visited:
                continue
            visited.add(neighbor)
            queue.append(neighbor)

    return order

File: Graph/php/bfs.php
<?php

declare(strict_types=1);

require_once __DIR__ . '/GraphUtils.php';

function breadthFirstTraversal(array $graph, string $source): array
{
    if ($source === '') {
        return [];
    }

    $order = [];
    $queue = [$source];
    $visited = [$source => true];

    while ($queue) {
        $current = array_shift($queue);
        $order[] = $current;

        foreach ($graph[$current] ?? [] as $neighbor) {
            if (!isset($visited[$neighbor])) {
                $visited[$neighbor] = true;
                $queue[] = $neighbor;
            }
        }
    }

    return $order;
}
